### ISSUE TITLE
[Performance] Flex Attention Block Mask Generation is very slow

### ISSUE THREAD

**Issue by Dogacel** [torchspec-project/TorchSpec#86]
One forward pass on Qwen3-8B takes around 350ms on H200 for a 4k sequence length. Looking at profiler data,

Optimizer step: 30ms
Backwards: 80ms
Forward: 220ms

Looking deeper into each forward step:
- Pre-norm: 2ms
- Flex attention: 4ms
- Flex attention create mask: 24ms
- MLP + norm: < 1ms
- Calculate loss: 3.5ms

This huge imbalance in flex attention mask dominates the training time and accounts for %70 of the forward pass time and if we somehow reduce this to < 10ms, it would cut down the training time by 50%.

<img width="855" height="346" alt="Image" src="https://github.com/user-attachments/assets/f6ff3475-8bca-4a0a-8fb0-cff0cf43c9db" />

**Comment by yubofredwang** [comment]
I think this should be solvable by padding to power of 2 sizes to avoid the causal mask building. We can also prebuild the mask cache. This is indeed something worth working on. Feel free to give it a try! I will also take a look this week
